cpp/nacjd.cc

```diff
@@ -0,0 +1,361 @@
+        json_new_titles << ',' << '\n';
+    json_new_titles << top_node->toString() << '\n';
+
+    return true;
+}
+
+
+void ExtractExistingIDsFromMarc(MARC::Reader * const marc_reader, std::set<std::string> &parsed_marc_ids) {
+    while (MARC::Record record = marc_reader->read()) {
+        std::string ppn(record.getControlNumber());
+        if (StringUtil::Contains(ppn, "[ICPSR]")) {
+            StringUtil::ReplaceString("[ICPSR]", "", &ppn);
+            StringUtil::TrimWhite(&ppn);
+            parsed_marc_ids.emplace(ppn);
+        }
+
+        std::string id_035(record.getFirstSubfieldValue("035", 'a'));
+        if (StringUtil::Contains(id_035, "[ICPSR]")) {
+            StringUtil::ReplaceString("[ICPSR]", "", &id_035);
+            StringUtil::TrimWhite(&id_035);
+            parsed_marc_ids.emplace(id_035);
+        }
+    }
+}
+
+
+void ExtractIDsFromWebsite(const std::set<std::string> &parsed_marc_ids, unsigned * const number_of_new_ids) {
+    const std::string DOWNLOAD_URL("https://www.icpsr.umich.edu/web/NACJD/search/studies?start=0&ARCHIVE=NACJD&PUBLISH_STATUS=PUBLISHED&sort=DATEUPDATED%20desc&rows=9000");
+    if (FileUtil::Exists(NACJD_TITLES))
+        FileUtil::DeleteFile(NACJD_TITLES);
+    
+    if (not Download(DOWNLOAD_URL, NACJD_TITLES, TIMEOUT_IN_SECONDS * 1000))
+        LOG_ERROR("Could not download website with nacjd ids.");
+    std::ifstream file(NACJD_TITLES);
+    if (file)
+        std::for_each(std::istream_iterator<char>(file),
+                      std::istream_iterator<char>(),
+                      HandleChar);
+    else
+        LOG_ERROR("couldn't open file: " + NACJD_TITLES);
+    
+    if (FileUtil::Exists(NACJD_NEW_TITLES_JSON) and not FileUtil::DeleteFile(NACJD_NEW_TITLES_JSON))
+        LOG_ERROR("Could not delete file: " + NACJD_NEW_TITLES_JSON);
+    std::ofstream json_new_titles(NACJD_NEW_TITLES_JSON);
+    json_new_titles << "{ \"nacjd\" : [ " << '\n';
+    bool first(true);
+    for (const auto &id : ids_website) {
+        if (parsed_marc_ids.contains(id))
+            continue;
+        bool success(DownloadID(json_new_titles, id, /*use_separator*/ not first));
+        if (first and success)
+            first = false;
+        if (success)
+            ++(*number_of_new_ids);
+    }
+    json_new_titles << " ] }";
+}
+
+
+void ParseJSONAndWriteMARC(MARC::Writer * const title_writer) {
+    std::string json_document;
+    FileUtil::ReadStringOrDie(NACJD_NEW_TITLES_JSON, &json_document);
+    JSON::Parser json_parser(json_document);
+    std::shared_ptr<JSON::JSONNode> internal_tree_root;
+    if (not json_parser.parse(&internal_tree_root))
+        LOG_ERROR("Could not properly parse \"" + NACJD_NEW_TITLES_JSON + ": " + json_parser.getErrorMessage());
+
+    const auto root_node(JSON::JSONNode::CastToObjectNodeOrDie("tree_root", internal_tree_root));
+    int no_total(0), no_title(0), no_description(0), no_license(0), no_initial_date(0), no_keywords(0), no_creators(0);
```

Comment:
Counters should always be unsigned because they can logically never be negative.